charmcraft/parts.py

```diff
@@ -257,6 +266,17 @@ def post_build_callback(self, step_info):
         """Collect metrics left by charm_builder.py."""
         instrum.merge_from(env.get_charm_builder_metrics_path())
 
+    def _get_os_special_path(self, path: Optional[str] = None) -> Optional[str]:
+        with suppress(OsReleaseIdError, OsReleaseVersionIdError):
+            os_release = os_utils.OsRelease()
+            if (os_release.id(), os_release.version_id()) in (("centos", "7"), ("rhel", "7")):
+                # CentOS 7 Python 3.8 from SCL repo
+                if path:
+                    return f"/opt/rh/rh-python38/root/usr/bin:{path}"
+                return f"/opt/rh/rh-python38/root/usr/bin:{os.environ.get('PATH', '')}"
```

Comment:
I don't think this is equivalent to the original code that you had in `get_build_environment()`; the returned dict should not use the value of the "current" environment variables, but placeholders like `${PATH}` (I think this is what you had originally) which are expanded later, when the build script is actually executed. Does that make sense?
From that point of view, it might be simpler to have this `_get_os_special_path` function receive no parameters and then have the callers append the correct strings.

Reply:
@cmatsuoka can you please check my reasoning here?

Reply:
Will this `PATH` only used in shell scripts? As you cannot set `PATH="${PATH}"` if it is not a shell.

Reply:
Yes, this dict is saved to a file that is then run with bash
https://github.com/canonical/craft-parts/blob/d70898380720fe37f6cce3d71d0919188fd79bac/craft_parts/plugins/rust_plugin.py#L119

Reply:
OK, so I had to leave `${PATH}` in the unit test. Since it is not call shell to run it.

Reply:
Since this one use `shlex.quote` later, it is not possible to use `${PATH}`. It get escaped to `env -i LANG=C.UTF-8 LC_ALL=C.UTF-8 'PATH=/opt/rh/rh-python38/root/usr/bin:${PATH}' SNAP=snap_value ...`. Unless I create that command by hand.

Reply:
For what it looks like in the test case results, `get_build_environment()` does not affect `get_build_commands()`.